CHIP-0011: BLS/SECP CLVM Operators and SOFTFORK Condition

danieljperry opened the pull request:
Opening PR for consideration of the BLS Additions CHIP.

hoffmang9 reviewed (dismissed):
lgtm

freddiecoleman reviewed (approved):
lgtm

hoffmang9 reviewed (dismissed):
lgtm. Shall I merge?

danieljperry commented:
No, let's keep it open while the CHIP is active. People can post their reviews as comments here.

danieljperry commented:
This video explains the details of this CHIP, including why we would like to add these operators and the technical details of how it would work:
https://www.youtube.com/watch?v=c4tKmjJEJCI
Reviews welcome!

danieljperry commented:
This CHIP has been updated to include a tentative soft fork block height. It will be moved to `Review` after the operators are confirmed included.

danieljperry commented:
This CHIP is now in `Review`. Note that a few of the operators have been added, removed, and modified since the CHIP was created. Significant changes are unlikely going forward. Please take another look at the CHIP's contents and leave your reviews here.

danieljperry commented:
The ability to verify secp256k1 and secp256r1 are now included with this CHIP. Please take a moment to review these additions and leave your thoughts here.

SlowestTimelord commented on CHIPs/chip-0011.md:
```diff
@@ -0,0 +1,411 @@
+
+Arguments: A G2 point followed by a list of G1/msg pairs
+
+Usage: `(bls_verify g1point1 msg1 g1point2 msg2 … g1pointn msgn)`
```
Is this usage example missing a G2 point as the first argument?

SlowestTimelord commented on CHIPs/chip-0011.md:
```diff
@@ -0,0 +1,411 @@
+
+Usage: `(secp256k1_verify pubkey msg_digest signature)`
+
+CLVM Cost: `850 000`
```
Why is k1 cost so much cheaper than r1? Are these costs benchmarked or from a cost breakdown like other opcodes above?

hoffmang9 replied:
Out of Bitcoin Core work, there have been very significant performance increases developed for k1 that have not occurred in r1. Hence this is a pretty empirical comparison of the two operators using the rust implementations of both.

SlowestTimelord commented on CHIPs/chip-0011.md:
```diff
@@ -0,0 +1,411 @@
+
+Functionality: Allows future conditions with non-zero CLVM costs to be added as soft forks. (This functionality was previously only possible as a hard fork)
+
+Arguments: The cost of the condition (specified in ten-thousands) is required as the first argument, further arguments are unspecified. The reason to scale the cost by `10 000` is to make the argument smaller. For example, a `cost` of `100` in this condition would equate to an actual cost of 1 million (`1 000 000`)
```
How does a low cost (e.g. `coinid`) get specified?

danieljperry replied:
Ten thousand is the minimum cost.

SlowestTimelord replied:
I'd like to understand the rationale for 10,000 specifically as the scaling factor. Why not 1,000 or 100,000? Are most op codes expected to be on the order of 10,000 cost? What is the impact of not being specific enough?

danieljperry replied:
To be clear, these are conditions. (opcodes mostly refer to operators in CLVM). The rationale to have a scaling factor at all is to save space (and cost for bytes) by not having to specify the trailing zeros in the full cost. The scaling factor could probably have been 100'000 or 200'000 as well. But conditions with costs less than 10'000 are not likely.

For comparison, the existing conditions with cost are `CREATE_COIN` at 1'800'000 and `AGG_SIG_*` at 1'200'000

SlowestTimelord commented on CHIPs/chip-0011.md:
```diff
@@ -0,0 +1,411 @@
+* [test-blspy-hash.txt](https://github.com/Chia-Network/clvm_rs/tree/main/op-tests/test-blspy-hash.txt)
+* [test-blspy-pairing.txt](https://github.com/Chia-Network/clvm_rs/tree/main/op-tests/test-blspy-pairing.txt)
+* [test-blspy-verify.txt](https://github.com/Chia-Network/clvm_rs/tree/main/op-tests/test-blspy-verify.txt)
+* [test-more-ops.txt](https://github.com/Chia-Network/clvm_rs/blob/main/op-tests/test-more-ops.txt)
```
Missing
https://github.com/Chia-Network/clvm_rs/blob/main/op-tests/test-secp256k1.txt
https://github.com/Chia-Network/clvm_rs/blob/main/op-tests/test-secp256r1.txt

danieljperry replied:
added, thanks

SlowestTimelord commented:
Reading about potential security concerns with how the secp256r1 constants were chosen, is there a particular reason to include it at all in addition to secp256k1?

hoffmang9 commented:
> Reading about potential security concerns with how the secp256r1 constants were chosen, is there a particular reason to include it at all in addition to secp256k1?

So there is certainly concern among some that the NSA may have poisoned r1's constants. I personally am dubious because it's the primary go to for Apple and Samsung and others in their Secure Elements. The "pro" here though is that r1 is NIST certified while k1 is not.

Choice of curves will remain fully optional for developers and end users also.

SlowestTimelord commented:
> So there is certainly concern among some that the NSA may have poisoned r1's constants. I personally am dubious because it's the primary go to for Apple and Samsung and others in their Secure Elements. The "pro" here though is that r1 is NIST certified while k1 is not.
> 
> Choice of curves will remain fully optional for developers and end users also.

Ah the industry use and NIST certification makes sense especially in the context of the May 2023 AMA. Thanks Gene.

SlowestTimelord commented:
> %: aka modulo, (Opcode: 60)
> modpow: computes (base ^ exponent) % modulus, (Opcode: 61)

The CHIP seems to indicates % as 61 and modpow as 60

danieljperry commented:
Sorry about that. I transposed them in the comment, which I have now fixed. The CHIP itself is correct.

danieljperry commented on CHIPs/chip-0011.md:
```diff
@@ -0,0 +1,453 @@
+
+Usage: `(% numerator denominator)`
+
+CLVM Cost: `998` base, `4` per byte (sum of both operands)
```
You are correct, good catch.

danieljperry commented on CHIPs/chip-0011.md:
```diff
@@ -0,0 +1,453 @@
+
+#### %
+
+Opcode: 61
```
> Is this just meant to be a simpler (and cheaper) version of what `divmod` already handles?

Correct, this operator only gives the remainder, and the cost is slightly less than that of `divmod`.

danieljperry replied:
> Any restrictions on negative numbers with `%`?

Negative numbers are allowed. I have added a clarifying note to the CHIP.

danieljperry commented:
This CHIP now includes the ability to add new CLVM conditions with cost. This functionality will become available after next year's hard fork. Currently, new conditions must have zero cost.
This CHIP also includes new AGG_SIG conditions, which only contain one or two of `parent coin`, `puzzle hash`, and `amount` in their signed messages (the current signature conditions contain all three). They will also include unique domain strings to prevent replay attacks.

Please review at your convenience.

danieljperry commented:
This CHIP is now in `Last Call`. If no further changes are required in the next two weeks (roughly), it will be moved to `Final`.

danieljperry commented:
The activation height for the fork described in this CHIP is now set to 4,474,000. Nothing else has been changed. This CHIP will remain in `Last Call` and the roughly two week window will be reset.

danieljperry commented:
We have to adjust the activation height for this CHIP once again, so I'm pulling it out of Last Call until we can determine a new height. Nothing else has changed with this CHIP.

danieljperry commented:
The soft fork activation height for this CHIP has been updated to 4,510,000 and the CHIP has been moved back into `Last Call`. If it does not need to be further modified for the next two weeks (roughly), then it will be moved to `Final`.

danieljperry commented:
This CHIP is now `Final`. No further changes are allowed.
